Create container-runtime-survey2021.md (#36)
supercontainers/containers-wg: _ideas/container-runtime-survey2021.md

# Container Runtime Survey 2021

> A survey to get some data on container runtime would be nice. Not only which once are used, but also what images are used (OCI) and how hardned the feelings are (use of multiple runtimes).

Reviewer: ```suggestion
A survey to get some data on container runtime would be nice. Not only which ones are used, but also what images are used (OCI) and how hardened the feelings are (use of multiple runtimes).
```

## Questions (early draft)

### Container Runtimes

| #  | Question | Type | Answers |
|:--:|:---------:|:----:|:---------------------:|
| R1 | How many HPC container runtimes are supported on the system you are working on? | single choice | 0,1,2,3,4,5 |

> | R2 | How many HPC container runtimes are you using in your daily live for the last 6 month? | single choice | 0,1,2,3,4,5 |

Reviewer: daily life == personal development and work development?
```suggestion
| R2 | How many HPC container runtimes are you using in your daily life for the last 6 month? | single choice | 0,1,2,3,4,5 |
```

> | R3 | Which HPC container runtimes are you using for your daily workloads? | multiple-choice | Charliecloud, Shifter, Singularity, Sarus, Podman |

Reviewer: How is this one different from the previous, is the difference just in the last 6 months or not? Also, we probably want to throw in Docker, it's fair game for using in personal development (I use it every day).

Reviewer: And should we have some ability to specify "Other" with a fill in the blank for it? E.g., udocker and some lesser known ones.

Author: I had Docker in before; I thought it is going to be hard to differentiate between a real HPC workload or an HPC system if we allow Docker to join.
On R1,R2,R3: Yeah, maybe R1&R2 with multiple choice plus other

Reviewer: Ah, that makes sense. Maybe in the description we can mention that we are intentionally not asking about Docker (so if they are looking for it and don't see it they don't exit the survey, which is something I've done before for other things, heh).

> | R4 | Which HPC container runtimes are on your radar but not yet used? | multiple-choice | Charliecloud, Shifter, Singularity, Sarus, Podman |

Reviewer: And other (fill in the blank)

### Images

| #  | Question | Type | Answers |
|:--:|:---------:|:----:|:---------------------:|

> | I1 | What description do you use regularily to build your images? | multi | Dockerfile, SIF recipe, ... |

Reviewer: We might want to account for recipe / other makers like hpc-container-maker, repo2docker, although those are primarily Dockerfile, it could be that the reader doesn't know that since it happens under the hood. Or just put another "Other" and fill in the blank here!

# Container Runtime Survey 2021

A survey to get some data on container runtime would be nice. Not only which ones are used, but also what images are used (OCI) and how hardened the feelings are (use of multiple runtimes).

## Questions (early draft)

### Container Runtimes

This section is focused to ask about HPC workloads running on HPC container runtimes. Some might use docker to run a single-node container on their workstation - we consider that out of scope because it is hard to differentiate between a service workloads and a HPC workload.

| #  | Question | Type | Answers |
|:--:|:---------:|:----:|:---------------------:|
| R1 | Which HPC container runtimes are supported on the system(s) you are working on? | multiple-choice | Charliecloud, Shifter, Singularity, Sarus, Podman, other |

> | R1.5 | Which HPC container runtimes are supported on the system(s) and not listed in R1? | text | |

Reviewer: Just to clarify - these X.5 questions are for the “other” choice? What does the 5 indicate?

Reviewer: I see the comment - thanks!

| R2 | Which HPC container runtimes are you using for your daily workloads? | multiple-choice | Charliecloud, Shifter, Singularity, Sarus, Podman, other |
| R2.5 | Which HPC container runtimes are you using for your daily workloads which are not listed in R2? | text | |
| R3 | Which HPC container runtimes are on your radar but not yet used? | multiple-choice | Charliecloud, Shifter, Singularity, Sarus, Podman, other |
| R3.5 | Which HPC container runtimes are on your radar but not yet used and not listed in R3? | text | |

### Images

| #  | Question | Type | Answers |
|:--:|:---------:|:----:|:---------------------:|
| I1 | What is the method you are using to build a container image from scratch? | multi | Dockerfile, SIF recipe, other |
| I1.5 | What method did we miss to list in I1? | text | |

> | I2 | Do you use higher-level abstractions on-top of the methods in I1? | multi | hpc-container-maker, repo2docker, spack, other |

Reviewer: Oh yeah good call on spack!

# Containers Working Group for HPC Survey

A survey to get some data on container runtime would be nice. Not only which ones are used, but also what images are used (OCI) and how hardened the feelings are (use of multiple runtimes).

> 

Reviewer 2: This text sounds ok for the actual survey, but here @kniec is trying to describe the idea/project first, IMO, right?

## Questions (early draft)

The `.5` questions are just asking about potential `other` answers we missed.

### About yourself

Just so that we can get a sense

| #  | Question | Type | Answers |
|:--:|:---------:|:----:|:---------------------:|

> | A1 | What is your primary environment? | single | National Lab, academia, commercial, other |

Reviewer 2: `s/National Lab/Govberment lab/` National lab is very US central, other places is just call gov lab

> | A2 | What is your function? | single | Scientist, Admin, RSE, other |

Reviewer 2: Agree with V
I would add the very basic `software engineer` (bc that's my job title)

> | A2.5 | Which functionis not listed in A2? | text | |

Reviewer: ```suggestion
| A2.5 | Which function is not listed in A2? | text | |
```

Reviewer 2: ++

> | A3 | How do you rate your experience with **non-HPC containers**? | single | beginner, intermediate, expert |

Reviewer 2: Very confusing question, you mean, web services and related topics?

> | A3 | How do you rate your experience with **non-HPC containers**? | single | beginner, intermediate, expert |

Author: I thought about that, yes. Experience with Dockerfile outside of the HPC context. How to create services... stuff like that.

> | A4 | How do you rate your experience with **HPC containers**? | single | beginner, intermediate, expert |

Reviewer: ```suggestion
| A4 | How do you rate your experience with **HPC containers**? | single | no experience, beginner, intermediate, expert |
```

Reviewer: Added this because I do think there is a difference between "I have never touched a container technology" and "I am a beginning with container technologies" and we would want to be able to tell these apart.

### Container Runtimes

> This section is focused to ask about HPC workloads running on HPC container runtimes. Some might use docker to run a single-node container on their workstation - we consider that out of scope because it is hard to differentiate between a service workloads and a HPC workload.

Reviewer: ```suggestion
This section is focused to ask about HPC workloads running on HPC container runtimes. While some might use docker to run a single-node container on their workstation, we do not consider it in this set because it is not traditionally used in HPC environment.
```

Reviewer: I agree that Docker is more "service" and the others possibly not, but I don't think the average user will make this distinction. If you look at the R community, for example, they run many pipelines with Docker and wouldn't label those runs as "services" and this question wording might be confusing.

Reviewer 2: the docker word could be replaced for generic `container tech` IMO

| #  | Question | Type | Answers |
|:--:|:---------:|:----:|:---------------------:|
| R1 | Which HPC container runtimes are supported on the system(s) you are working on? | multiple | Charliecloud, Shifter, Singularity, Sarus, Podman, other |
| R1.5 | Which HPC container runtimes are supported on the system(s) and not listed in R1? | text | |
| R2 | Which HPC container runtimes are you using for your daily workloads? | multiple | Charliecloud, Shifter, Singularity, Sarus, Podman, other |
| R2.5 | Which HPC container runtimes are you using for your daily workloads which are not listed in R2? | text | |
| R3 | Which HPC container runtimes are on your radar but not yet used? | multiple | Charliecloud, Shifter, Singularity, Sarus, Podman, other |
| R3.5 | Which HPC container runtimes are on your radar but not yet used and not listed in R3? | text | |

> | R4 | Do you aware of the OCI runtime spec? | single choice | yes / no |

Reviewer 2: ```suggestion
| R4 | Are you aware of the OCI runtime spec? | single choice | yes / no |
```

Reviewer 2: that or `Do you acknowledge` , but as it is is not well redacted

Author: jeezes christ

| R5 | Do you care about the OCI runtime spec? | single choice | yes / no |

### Images

> The image section aims to get an overview of how folks are dealing with images

Reviewer 2: ```suggestion
The image section aims to get an overview of how users are interacting with container images (building, sharing, Storing)
```

| #  | Question | Type | Answers |
|:--:|:---------:|:----:|:---------------------:|

> | I1 | What is the method you are using to build a container image from scratch? | multi | Dockerfile, SIF recipe, other |

Reviewer 2: ```suggestion
| I1 | What is the method you are using to build a container image? | multi | Dockerfile, SIF recipe, From scratch, other |
```

Reviewer 2: From scratch is an actual valid way supported by Buildah

Reviewer 2: now that I think it better this question is wrong, you `build` a container with a container runtime tool, a Dockerfile or SIF def, is a way to `DEFINE` a  container image. we should have 2 questions
1 preferred def file 
2 way of building the def file

| I1.5 | What method did we miss to list in I1? | text | |
| I2 | Do you use higher-level abstractions on-top of the methods in I1? | multi | hpc-container-maker, repo2docker, spack, other |
| I2.5 | What method did we miss to list in I2? | text | |

> | I3 | Once build, do you push the image to a central repository? | single | yes / no |

Reviewer: ```suggestion
| I3 | Once build, do you push the image to a central registry? | single | yes / no |
```

> | I4 | What container repositories are you pushing builds to? | multi | dockerhub, gcr, ecr, quay, github, azure |

Reviewer: ```suggestion
| I4 | What container registries are you pushing builds to? | multi | Docker Hub, Google Container Registry (gcr), Elastic Container Registry (ecr), Quay.io, GitHub Packages (ghcr,io), Azure |
```

Reviewer 2: I would re-write I3 and I4, some labs use NFS systems to store the container images, so they don't `PUSH` , simple move.
so I3 should be, `How do you store your container images after build`
I4 Which container registry is used at your institution to store and share images

that way answers will be more meaning full to reviewers

### Potentials

Without making the survey too long, we could also ask the following questions.

1. do you deal with different target architectures in your build process?

> 2. Do you use CI/CD for automated build and deploy?

Reviewer: ```suggestion
2. Do you use CI/CD for automated build and deploy?
3. What are you biggest challenges or pain points when using containers, or reasons that you don't use them (fill in the blank)
4. What do containers not do that you wish they could? What would be an ideal future?
```

Reviewer 2: I like the intend but when reviewing the results I would like to have more insight
2. Does you institution has a CI implemented for container image builds?
3. Does your institution allows you to pull from external registries without going through the institution CI system?
4. How do you utilize built container images from CI (if any) from the main/production system?
5. Does your institution containerized system allows you to define automated deployments, if yes, how?
4. What are you biggest challenges or pain points when using containers, or reasons that you don't use them (fill in the blank)
5.  What do containers not do that you wish they could? What would be an ideal future?

Author: How about a CI/CD section in the survey?

Reviewer: I have mixed feelings about this, because asking about CI is out of scope for what container technology is being used, which I think was the focus of this survey. However, if this is considered relevant or important, I could see one question being added. Maybe something like:

**What do you use to build and deploy containers?**
- GitHub Workflows
- Automated builds via a registry (e.g., Docker Hub)
- On premise resources
- Circle CI
- Travis CI

I guess I'm not too keen on the above because we are basically asking about CI services. Maybe it's more appropriate to do:

**What do you use to build and deploy containers?**
- A provided CI service (GitHub Workflows, Travis CI, Circle CI, or other)
- Automated builds via a registry (e.g., Docker Hub)
- On premise resources

And then any major categories I'm missing. I'm not sure we want to (yet) be asking about rules/permissions at an org, that goes a little into privacy more than I think this survey should (e.g., to possible make it easier to identify the respondent). And the question also assumes that an org has an "institutional CI system" (most in my experience outside of national labs and industry do not). I also don't think the average user is going to know much about defining automated deployments, and @ArangoGutierrez see below - I added already the questions about pain points and ideal futures :)

Reviewer 2: My mixed feeling here is `What do you use to build and deploy containers?` in Academia and/or HPC people are not used to the `deploy containers` wording, and I wouldn't expect any HPC institution to deploy from GitHub actions. hence why  I think the questions should be done separately, a set of questions for builds and other for testing and deployment. Again, the question `What do you use to build and deploy containers?` won't provide insightful data.
If we want insightful data we need to ask about these topics by separate, or we will end up with useless data and nice graphs. 
50% use github actions.... and what? what does that tell us?
20% has on premise build and deploy.... oh awesome, what could that be.... 

We need questions that will lead to useful and decision making data, so we as a community can better direct the effort on future feature development. That is why I am advocating for very targeted questions, so then developers of Singularity/Podman/CharliCloud can sit down and say, oh users want more of X, or User are really not using Y let's not invest so many devel hours on that feature.

Or For people like Chris, knowing how people really work, and which tooling they use, can help him push ideas and features inside his org, and other could do the same for other clouds.

Reviewer: > My mixed feeling here is What do you use to build and deploy containers? in Academia and/or HPC people are not used to the deploy containers wording

Disagree - my labs / user groups at Stanford did this, and we do at the lab too.

> and I wouldn't expect any HPC institution to deploy from GitHub actions. 

Also disagree, my institution does.

> We need questions that will lead to useful and decision making data, so we as a community can better direct the effort on future feature development.

Agree partially. This first round at the survey I think should be generally. I think it would be okay to have separate questions for build and deploy, but I think most people are probably building and deploying from the same resource, so it's not so cray to group them together. I think we should start with a conservative strategy and only add _more_ questions when it's decided that is a direction the survey should go in. If we start with too many questions, period, people won't take it.

Reviewer 2: What does Deploy means in this discussion, do you allow GitHub Actions to deploy in production clusters? like running jobs against SLURM?  or Deploy as pushing the image to the cluster? The "deploy" word is used in diff ways in academia and cloud.

Reviewer 2: Most real life Prod clusters have some sort of air gap, or at least VPN or proxy protection, + running the final job with a queue manager sometimes require some perms, that is my concept of "deploy"

Author: I am on the conservative and short side here. As a first stab I reckon we learn already something even if it is 
* 20% use docker build interactively
* 30% use their own scripting to speed up repetitive tasks
* 50% use CI/CD driven by a artefact repository (if only)

Author: That said, I might speak out of the desperation that we do not have any data now. :) 
I am after the basic stuff as a first go.

Reviewer 2: This conversation was not resolved before merging the PR @vsoch @kniec , why?

Reviewer: It was per @kniec request in the other PR, @ArangoGutierrez why don’t you create a new issue with unresolved concerns so we can pick up the discussion. The threads here were getting unruly and I think @kniec just didn’t see them. I also think it’s important we have a rough draft merged so I can start on paperwork because these nits can go on a very long time!

| R2.5 | Which HPC container runtimes are you using for your daily workloads which are not listed in R2? | text | |
| R3 | Which HPC container runtimes are on your radar but not yet used? | multiple | Charliecloud, Shifter, Singularity, Sarus, Podman, other |
| R3.5 | Which HPC container runtimes are on your radar but not yet used and not listed in R3? | text | |
| R4 | Do you aware of the OCI runtime spec? | single choice | yes / no |
| R5 | Do you care about the OCI runtime spec? | single choice | yes / no |

### Images

The image section aims to get an overview of how folks are dealing with images

| #  | Question | Type | Answers |
|:--:|:---------:|:----:|:---------------------:|
| I1 | What is the method you are using to build a container image from scratch? | multi | Dockerfile, SIF recipe, other |
| I1.5 | What method did we miss to list in I1? | text | |
| I2 | Do you use higher-level abstractions on-top of the methods in I1? | multi | hpc-container-maker, repo2docker, spack, other |
| I2.5 | What method did we miss to list in I2? | text | |
| I3 | Once build, do you push the image to a central repository? | single | yes / no |

> | I4 | What container repositories are you pushing builds to? | multi | dockerhub, gcr, ecr, quay, github |

Reviewer: Does Microsoft have one? https://azure.microsoft.com/en-us/services/container-registry/
```suggestion
| I4 | What container repositories are you pushing builds to? | multi | dockerhub, gcr, ecr, quay, github, azure |
```

### Potentials

> Without making the survey to long, we could also ask about.

Reviewer: ```suggestion
Without making the survey too long, we could also ask the following questions.
```

> 1. do you deal with different target architectures in your build process?

Reviewer: Do we care which architectures?
Maybe something about “I need my container to work with” … GPU, MPI, etc.

Author: > Do you need your container to be tailored for a specific system (MPI, GPU)?
But I am not sure if we want to make the survey go down this complexity route which we do not yet have a common understanding for...

Reviewer: Yeah that’s fair!

Reviewer 3: I do care about alternative architectures. However I would like to suggest question should be more general to include GPUs and/or accelerators, rather than just CPU ISAs. 

"In your container build process, do you target more than 1 type of hardware configuration (different CPUs, GPUs, accelerators, ...)?"

Reviewer 2: I would call it `build process` , the question should be
> Do you have to build your container image for different CPU architectures?
> DO you care about available accelerators(GPU, FPGA, specialized network hardware) when building your container image?

Reviewer: I'm with @kniec on this one - if we ask if they care about different architectures, we could probably guess the handful that are relevant. I thin asking to provide more information here won't get us much further, and it would make the survey longer and possibly lose a few folks.

Reviewer 2: The question it self is bad redacted, you don't "deal" with multiple archs, your institute owns diff arch clusters, or you "want" to test your binary over diff CPU's. but users, don't "deal" with multiple CPU's during build. that is why I prefer `Do you have to build your container image for different CPU architectures?` and or `DO you care about available accelerators(GPU, FPGA, specialized network hardware) when building your container image?` , the question as it is, won't give the data reviewers any useful insightful data.

Reviewer: Ah okay I misunderstood, - you are saying to have _two_ questions, one about architectures (very similar to what we have plus CPU) and then a _different one_ for accelerators.  You said "the question should be" so I missed that you were suggesting multiple. 

I would be fine with that.

> 2. Do you use CI/CD?

Reviewer: ```suggestion
2. Do you use CI/CD for automated build and deploy?
```
We might want to distinguish using it for test, build and deploy, or both.
